spec/history.cddl

```diff
@@ -0,0 +1,58 @@
+can-history = tagged<{
+  total_num_changes : unsigned
+  changes : [* can-change]
+}>
+
+can-change = {
+  timestamp : timestamp
+  origin : can-change-origin
+  type : can-change-type
+}
+
+can-change-origin =
+  from-user /
+  from-canister
+
+from-user = {
+  user_id : principal
+}
+
+from-canister = {
+  canister_id : principal
+  canister_history_index : unsigned
+}
+
+can-change-type =
+  canister-creation /
+  code-uninstall /
+  code-deployment /
+  controllers-change
+
+canister-creation = 0
+code-uninstall = 1
+
+code-deployment = {
+  mode : code-deployment-mode
+  module_hash : bytes
+}
```

Comment:
This looks strangly non-uniform. Some fields are numbers, some are records with a `mode`.

Moreover, it’s not really a `can-change-type` (NB: type) if it also has additional data! Why not `can-change`?

Or, maybe more idiomatic, one could simply inline the fields into the `can-change` type.